sled-agent/src/bootstrap/trust_quorum/server.rs

```diff
@@ -0,0 +1,128 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::io;
+use std::net::{Ipv6Addr, SocketAddr, SocketAddrV6};
+
+use slog::Logger;
+use tokio::net::{TcpListener, TcpStream};
+use tokio::task::JoinHandle;
+use vsss_rs::Share;
+
+use super::msgs::Response;
+use crate::bootstrap::{agent::BootstrapError, spdm};
+
+// TODO: Get port from config
+// TODO: Get IpAddr from local router:
+//   See https://github.com/oxidecomputer/omicron/issues/443
+pub const PORT: u16 = 12347;
+
+/// A TCP server over which a secure SPDM channel will be established and an
+/// application level trust protocol will run.
+pub struct Server {
+    log: Logger,
+    share: Share,
+    listener: TcpListener,
+}
+
+impl Server {
+    pub fn new(log: &Logger, share: Share) -> io::Result<Self> {
+        let addr = SocketAddrV6::new(Ipv6Addr::UNSPECIFIED, PORT, 0, 0);
+        let sock = socket2::Socket::new(
+            socket2::Domain::IPV6,
+            socket2::Type::STREAM,
+            Some(socket2::Protocol::TCP),
+        )?;
+        sock.set_only_v6(true)?;
+
+        // Allow rebinding during TIME_WAIT
+        sock.set_reuse_address(true)?;
+
+        sock.bind(&addr.into())?;
+        sock.listen(5)?;
```

Comment:
The only way to get a `tokio::TcpListener` instance is to call bind, which then should be awaited immediately. I wrote things this way so that new would not have to be `async` and would allow me to construct a `Server` without any async code. I'm not sure this is 100% necessary, and I can certainly change the API, but it seems pretty inconsequential.

As for the `5`, habbit from 20 years of the Stevens book?  It was never clear to me what the queue backlog should be.

Reply:
Isn't this being handed off to a `tokio::net::TcpListener` anyway?

I mentioned it because, as an example (from the tokio docs), it seems like using a `socket2` (or `std::net`) socket under the hood shouldn't require an explicit call to `listen()`, right?

```rust
use std::error::Error;
use tokio::net::TcpListener;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let std_listener = std::net::TcpListener::bind("127.0.0.1:0")?;
    std_listener.set_nonblocking(true)?;
    let listener = TcpListener::from_std(std_listener)?;
    Ok(())
}
```

Reply:
I just tried removing the listen call and the test times out. I'd prefer to leave this as is for now. It uses all the fancy setup so we can get IPv6 only traffic. We can change this when we have our own prefix, and no longer need to use UNSPECIFIED.